src/sc/sidechaintypes.h

```diff
@@ -10,15 +10,47 @@
 #include <boost/variant.hpp>
 
 #include <zendoo/zendoo_mc.h>
-#include <zendoo/error.h>
 
 #include "uint256.h"
 #include "hash.h"
 #include "script/script.h"
 #include "amount.h"
 #include "serialize.h"
-///////////////////////////////// Field types //////////////////////////////////
 
+class CZendooCctpObject
+{
+public:
+	CZendooCctpObject() = default;
+    virtual ~CZendooCctpObject() = default;
+
+    CZendooCctpObject(const std::vector<unsigned char>& byteArrayIn): byteVector(byteArrayIn) {}
+    virtual void SetByteArray(const std::vector<unsigned char>& byteArrayIn) = 0; //Does custom-size check
+    const std::vector<unsigned char>& GetByteArray() const;
+
+    void SetNull();
+    bool IsNull() const;
+
+    virtual bool IsValid() const = 0;
+    bool operator==(const CZendooCctpObject& rhs) const { return this->byteVector == rhs.byteVector; }
+    bool operator!=(const CZendooCctpObject& rhs) { return !(*this == rhs); }
+
+    // SERIALIZATION SECTION
+    ADD_SERIALIZE_METHODS
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(byteVector);
+        //Allow zero-length byteVector for, e.g. serialization/hashing of empty txes/certs
+        if (byteVector.size() != 0 && byteVector.size() != SerializedSize())
```

Comment:
I'm worried about using byte vector to serialize a constant size value. For example, for FieldElement we always know the size. So by default, uninitialized/invalid FE can be set to the static value of `FF...FF` - which is invalid FE.
We may make FE backed by base_blob of given size that will save us some bytes during serialization: no `VarInt` serialization for byte array length.
Current approach, as for me, contradicts with the regular serialization approach and with how we define an "uninitialized" state: setting empty byte vector is like setting a different invalid type instead of setting an invalid value specific for the type (-1 for Amount, or "FF...FF" for FE).
Maybe I'm missing something, but hope it's useful.

Reply:
Instead of throwing a runtime exception for not having deserialized the expected number of bytes, also considered that both proof and vk will have a variable size, remove the check about the expected size and move it to the concrete class IsValid method.
Remove also the SerializedSize methods from derived classes